### ISSUE TITLE
Error: object '.SDcols' not found

### ISSUE THREAD

**Issue by drag05** [Rdatatable/data.table#5742]
I have encountered this issue for the first time:

```
require(data.table)
tt = as.data.table(iris)


A. tt[, .SD, .SDcols != is.character]

Error in eval(bysub, x, parent.frame()) : object '.SDcols' not found

B. head(tt[, .SD, .SDcols != is.character, with = FALSE])

Null data.table (0 rows and 0 cols)

C. head(tt[, (.SDcols) := sapply(.SD, is.factor), .SDcols = is.factor])

Error in eval(lhs, parent.frame(), parent.frame()) : 
  object '.SDcols' not found

D. head(tt[, (.SDcols) := sapply(.SD, is.character), .SDcols = is.factor])

Error in eval(lhs, parent.frame(), parent.frame()) : 
  object '.SDcols' not found





sessionInfo()

R version 4.3.2 (2023-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

attached base packages: [1] stats     graphics  grDevices utils     datasets  methods   base      
other attached packages: [1] data.table_1.14.8  loaded via a namespace (and not attached): 
compiler_4.3.2    tools_4.3.2       rstudioapi_0.15.0
--
 
Thank you!

**Comment by MichaelChirico** [comment]
You probably meant `.SDcols = !is.character`.

**Comment by jangorecki** [comment]
`.SDcols` on the LHS of `:=` is not yet implemented.

**Comment by drag05** [comment]
> You probably meant `.SDcols = !is.character`.

```
head(tt[, .SD, .SDcols = !is.factor])

   Sepal.Length Sepal.Width Petal.Length Petal.Width
1:          5.1         3.5          1.4         0.2
2:          4.9         3.0          1.4         0.2
3:          4.7         3.2          1.3         0.2
4:          4.6         3.1          1.5         0.2
5:          5.0         3.6          1.4         0.2
6:          5.4         3.9          1.7         0.4
 ```
Why the difference between != is.factor and = !is.factor ?

**Comment by jangorecki** [comment]
`!= ...` is evaluated in parent scope before `[` really starts, `= !...` is evaluated from inside `[`

**Comment by drag05** [comment]
@jangorecki  Well, this is a fine distinction and I assume is not the only case. 

Maybe these cases could be (if not have been already!) exemplified under one Section in the documentation, together with equivalent use of  `eval(..., envir)`  or `eval(parse( ...))` by case.

Thank you!

**Comment by drag05** [comment]
@jangorecki  I will post examples of eval(parse) once I find some time for it.
Until then, I am giving this example from `data.table` special symbols documentation which, I find curious :

```
DT = data.table(x=rep(c("b","a","c"),each=3), v=c(1,1,1,2,2,1,1,2,2), y=c(1,3,6), a=1:9, b=9:1)
DT
   x v y a b
1: b 1 1 1 9
2: b 1 3 2 8
3: b 1 6 3 7
4: a 2 1 4 6
5: a 2 3 5 5
6: a 1 6 6 4
7: c 1 1 7 3
8: c 2 3 8 2
9: c 2 6 9 1
```
If `.SDcols` represents the column names of `.SD` and I want to obtain a data table with all columns of DT except column "v" I would be inclined to do this:

```
DT[, .SD, .SDcols != 'v']  # Error in eval(bysub, x, parent.frame()) : object '.SDcols' not found
```
The correct way is
```
DT[, .SD, .SDcols = -'v']  # correct
   x y a b
1: b 1 1 9
2: b 3 2 8
3: b 6 3 7
4: a 1 4 6
5: a 3 5 5
6: a 6 6 4
7: c 1 7 3
8: c 3 8 2
9: c 6 9 1
```
But not
```
DT[, .SD, .SDcols = - v] # Error in eval(colsub, parent.frame(), parent.frame()) : 
                                       object 'v' not found
```
However, If I only want column v:

```
DT[, .SD, .SDcols = 'v']  # correct
   v
1: 1
2: 1
3: 1
4: 2
5: 2
6: 1
7: 1
8: 2
9: 2
```
But not
```
DT[, .SD, .SDcols = v]  # Error in eval(colsub, parent.frame(), parent.frame()) : 
                                     object 'v' not found

```
I find this last error confusing since it represents a particular case of

```
DT[, .SD, .SDcols = v:b] 

   v y a b
1: 1 1 1 9
2: 1 3 2 8
3: 1 6 3 7
4: 2 1 4 6
5: 2 3 5 5
6: 1 6 6 4
7: 1 1 7 3
8: 2 3 8 2
9: 2 6 9 1
```
And similar for 
```
DT[, .SD, .SDcols = -(v:b)] 
   x
1: b
2: b
3: b
4: a
5: a
6: a
7: c
8: c
9: c
```
So, it cannot find one element of the set v:b in parent frame however, it is capable to find the entire set v:b;  `.SDcols` takes 'v' as column name but not v. At the same time `.SDcols` takes v:b as column names but not any of them alone, unless I write it like this:
```
DT[, .SD, .SDcols = v:v] 
   v
1: 1
2: 1
3: 1
4: 2
5: 2
6: 1
7: 1
8: 2
9: 2
```
In conclusion, `.SDcols` only accepts assignment `=`. However the RHS should be quoted or not quoted depending whether is one element or a list. Why? 
Thank you!

**Comment by drag05** [comment]
@jangorecki Never stated `.SDcols` represent all columns of DT. 
I always use the combination `.SD` / `.SDcols` to select columns in DT as I find it the most trustful so far - had some cases when using `..(names)` in j threw error.  I am OK with using sometimes quoted and sometimes unquoted however, I would like to have a documented explanation/rules about this.
Thank you!

**Comment by jangorecki** [comment]
Is the current documentation of `.SDcols` in
```r
?data.table
```
not sufficient enough?

at least in development version it feels it covers well those use cases

**Comment by drag05** [comment]
@jangorecki  The documentation on page 8 in Manual is sufficient as procedure (what) but less clear as why.  Maybe there are some other resources of which I am not aware?
Thank you!

**Comment by jangorecki** [comment]
I believe it was because of popular demand.
The only way to figure that out now (as it was growing organically) is to git blame documentation entry for `.SDcols` and see which issues have been resolved by each commits changing documentation.

**Comment by drag05** [comment]
@jangorecki Regarding the error  "in eval(bysub, x, parent.frame()) : object '.SDcols' not found":
shouldn't lexical scoping look first in current environment then up toward global environment when an object is not found? The error suggests an inverse search path: starting in parent frame and not moving down to current environment.

**Comment by jangorecki** [comment]
`eval` call defines where it is being looked for, first in `x`, then in `parent.frame()`
```r
f = function(x, i, j, by) {
  bysub = substitute(by)
  eval(bysub, x, parent.frame())
}
f(list(), , , .SDcols!=1)
#Error in eval(bysub, x, parent.frame()) : object '.SDcols' not found

.SDcols = 1
f(list(), , , .SDcols!=1)
#[1] FALSE
f(list(.SDcols=0), , , .SDcols!=1)
#[1] TRUE
```

**Comment by drag05** [comment]
@jangorecki So, if, in parent.frame(), I set `.SDcols = names(DT)`  then, I will be able to select columns in DT using `!=`,  or maybe `%in%` as follows:

DT = data.table(grp=rep(c("A", "B", "C", "A", "B"), c(2,2,3,1,2)), value=1:10)
DT
   grp value
 1:   A     1
 2:   A     2
 3:   B     3
 4:   B     4
 5:   C     5
 6:   C     6
 7:   C     7
 8:   A     8
 9:   B     9
10:   B    10

.SDcols = names(DT)

DT[, .SD, .SDcols != grp]
    .SDcols value
 1:    TRUE     1
 2:    TRUE     2
 3:    TRUE     3
 4:    TRUE     4
 5:    TRUE     5
 6:    TRUE     6
 7:    TRUE     7
 8:    TRUE     8
 9:    TRUE     9
10:    TRUE    10
```
Or, with a bit of overhead

```
DT[, .SD, .SDcols != grp][, .SDcols := NULL][]

   value
 1:     1
 2:     2
 3:     3
 4:     4
 5:     5
 6:     6
 7:     7
 8:     8
 9:     9
10:    10
```
with no quotes and equivalent to:

```
DT[, .SD, .SDcols = !'grp']
    value
 1:     1
 2:     2
 3:     3
 4:     4
 5:     5
 6:     6
 7:     7
 8:     8
 9:     9
10:    10
```
with quotes.

So, why direct `eval()` in `f` to look in `parent.frame()` instead the current environment in the first place?  

```
.SDcols = 'a'

 f = function(x, i, j, by) {
     bysub = substitute(by)
     eval(bysub, x)
 }

 f(list(), , , .SDcols != 1)
[1] TRUE

f(list(), , , .SDcols != 'a')
[1] FALSE
```

**Comment by jangorecki** [comment]
You need to understand that when you use != you are not supplying .SDcols anymore. You are supplying an argument matched by position, not .SDcols, and you are supplying logical vector to it. Then it doesn't matter if you call your variable .SDcols or .AnythingElse.
`eval` scoping current environment would be a bug because you could refer to `bysub`. parent.frame() is used as enclosing environment only, in the first place it looks in `x`.

**Comment by drag05** [comment]
@jangorecki  I have little understanding of what .SDcols is and have some intuition on how it works in scripts. 
Thank you for your time!
